README.md
<h1>Documentation Plugin Home Center de Fibaro pour SARAH</h1><br />
<p>
Auteur : e1000 (code inspiré du plugin Zibase de Baptiste MARIENVAL)<br />

<p><i>Version 0.3.2 (26/01/2016)</i></p>

1. CONFIGURATION : homecenter.prop<br />
2. MODULES : devices.xml<br />
3. ACTIONS : homecenter.xml<br />
4. COMMENT ÇA MARCHE ?<br />

<h2>1. CONFIGURATION : <tt>homecenter.prop</tt></h2>

<p>Ce fichier contient les éléments indispensables pour la connexion à votre box Home Center de Fibaro :</p>

<ul>
<li><b>url</b></li>
<p>Adresse IP ou nom de domaine sur lequel votre Home Center est accessible (ex.: 192.168.1.12).<br />
Pour connaître l'adresse IP de votre box, rendez-vous dans la partie <strong>Configuration</strong> puis <strong>Paramètres LAN</strong>.</p>
<img style="width:1000px;" src="./plugins/homecenter/images/configuration_reseau.png" />

<li><b>port</b></li>
<p>Port utilisé par le serveur de la box Home Center. Par défaut, le port 80 est utilisé.</p>

<li><b>username</b></li>
<p>Identifiant de l'utilisateur utilisé pour contrôle votre Home Center à distance.
Il est conseillé de créer un compte dédié à cet usage, et de ne pas utiliser le compte Admin.</p>

<li><b>password</b></li>
<p>Le mot de passe correspondant au compte ci-dessus.</p>
</ul>

<h2>2. MODULES : <tt>devices.xml</tt></h2>

<p>Tous les périphériques que vous souhaitez contrôler via SARAH sont définis dans le fichier <tt>devices.xml</tt>.</p>

<p>La configuration se veut la plus simple possible et ne nécessite que 4 types d'informations.<br />
À terme, la génération de ce fichier pourra être faite automatiquement.</p>
<p>Exemple :</p>
<pre>
&lt;devices&gt;
	&lt;!-- SONDES --&gt;
	&lt;device name="temp_salon" 		type="com.fibaro.temperatureSensor"	id="14" 	tts="La température du salon est de %s degrés " /&gt;
	&lt;device name="hygro_sdb" 		type="com.fibaro.humiditySensor" 	id="55" 	tts="L'hygrométrie de la salle de bain est de %s pourcent " /&gt;
	&lt;device name="lumi_entree" 		type="com.fibaro.lightSensor" 		id="132" 	tts="La luminosité de l'entrée est de %s pourcent " /&gt;
	&lt;!-- DETECTEURS --&gt;
	&lt;device name="mouvement_sam" 		type="com.fibaro.motionSensor" 		id="9" 		tts="Mouvement dans la salle à manger " /&gt;
	&lt;device name="ouverture_2" 		type="com.fibaro.doorSensor" 		id="96" 	tts="Capteur de porte 2 ouvert " /&gt;
	&lt;device name="fumee_chaufferie" 	type="com.fibaro.FGSS001" 		id="218" 	tts="Il y a de la fumée dans la chaufferie " /&gt;
	&lt;device name="chaleur_combles" 		type="com.fibaro.heatDetector" 		id="227" 	tts="Température anormale dans les combles " /&gt;
	&lt;device name="inondation_buanderie" 	type="com.fibaro.FGFS101" 		id="186"	tts="Inondation dans la buanderie " /&gt;
	&lt;!-- com.fibaro.binarySwitch --&gt;
	&lt;device name="lampe_cuisine" 		type="com.fibaro.binarySwitch" 		id="39" 	tts=" la lampe de la cuisine " /&gt;
	&lt;device name="machine_a_laver" 		type="com.fibaro.binarySwitch" 		id="158" 	tts=" la machine à laver " /&gt;
	&lt;!-- com.fibaro.multiLevelSwitch --&gt;
	&lt;device name="lampe_chambre" 		type="com.fibaro.multiLevelSwitch" 	id="80" 	tts=" la lampe de la chambre " /&gt;
	&lt;!-- com.fibaro.blind --&gt;
	&lt;device name="volet_chambre" 		type="com.fibaro.blind	 		id="80" 	tts=" le volet de la chambre " /&gt;
	&lt;!-- VIRTUELS --&gt;
	&lt;device name="volet_tous" 		type="virtual_device" 			id="102" 	tts=" tous les volets "		buttonOn="1" buttonMy="2" buttonOff="3" slider="4" /&gt;
	&lt;device name="lampe_escalier" 		type="virtual_device" 			id="69" 	tts=" la lampe de l'escalier "	buttonOn="1" buttonOff="2" /&gt;
&lt;/devices&gt;
</pre>

Pour chaque élément <tt>&lt;device ... /&gt;</tt>, ont trouve :
<ul>
	<li><b>name</b></li> 
	<p><b>Nom du module</b> qui doit correspondre au nom utilisé pour les actions associées définies dans le fichier <tt>homecenter.xml</tt> décrit plus bas.<br />
	Ce nom doit être unique, mais ne doit pas obligatoirement correspondre au nom enregistré dans l'interface Home Center.<p>
	<li><b>type</b></li> 
	<p><b>Type de module</b> tel qu'ils sont définis dans l'API Fibaro. Il s'agit bien du <tt>type</tt> et non du <tt>basetype</tt>. <br />
	Ce paramètre est utilisé pour connaître les actions possibles sur ce module.
	</p>
	<li><b>id</b></li>
	<p>L'<b>identifiant</b> Home Center du module.</p>
	<li><b>tts</b></li>
	<p>Le <b>texte</b> que doit prononcer SARAH en retour lorsque l'action est réalisée. Pour les sondes, la chaine <tt>%s</tt> sera remplacée par la valeur retournée.</p>
</ul>
Dans le cas des modules virtuels uniquement, on trouve des paramètres supplémentaires :
<ul>
	<li><b>buttonOn</b></li>	
	<p>Ce paramètre précise le numéro du bouton à presser pour activer le module via la commande <tt>pressButton</tt>
	en remplacement de la commande <tt>turnOn</tt>.
	<li><b>buttonOff</b></li>	
	<p>Ce paramètre précise le numéro du bouton à presser pour désactiver le module via la commande <tt>pressButton</tt>
	en remplacement de la commande <tt>turnOff</tt>.
	<li><b>buttonMy</b></li>	
	<p>Spécifique aux volets roulants Somfy, ce paramètre précise le numéro du bouton à presser pour mettre le volet en position « My ».
	<li><b>slider</b></li>	
	<p>Ce paramètre précise le numéro du slider permettant de régler le module via la commande <tt>setSlider</tt>
	en remplacement de la commande <tt>setValue</tt>.
</ul>

<p>Voici la liste (non exhaustive) des types de modules :</p>
<ul>
<li>Capteurs ou sondes :</li>
<dl>
	<dt>com.fibaro.multilevelSensor</dt>
	<dd>Capteur générique</dd>
	<dt>com.fibaro.lightSensor</dt>
	<dd>Capteur de luminosité</dd>
	<dt>com.fibaro.temperatureSensor</dt>
	<dd>Capteur de température</dd>
	<dt>com.fibaro.humiditySensor</dt>
	<dd>Capteur d'humidité</dd>
</dl>
<li>Actionneurs :</li>
<dl>
	<dt>com.fibaro.binarySwitch</dt>
	<dd>Actionneur tout ou rien (On/Off)</dd>
	<dt>com.fibaro.multilevelSwitch</dt>
	<dd>Gradateur (ou <i>Dimmer</i>)</dd>
	<dt>com.fibaro.blind</dt>
	<dd>Volet roulant</dd>
</dl>
<li>Détecteurs de sécurité :</li>
<dl>
	<dt>com.fibaro.motionSensor</dt>
	<dd>Détecteur de mouvement. Les périphériques esclaves de certains modules sont de ce type (exemple : capteur d'inondation FGFS101)</dd>
	<dt>com.fibaro.FGFS101</dt>
	<dd>Détecteur d'inondation</dd>
	<dt>com.fibaro.FGSS001</dt>
	<dd>Détecteur de fumée ; à noter que les détecteurs de deuxième génération (FGSD) sont également identifiés sous ce type</dd>
	<dt>com.fibaro.heatDetector</dt>
	<dd>Détecteur de chaleur anormalement élevée</dd>
	<dt>com.fibaro.doorSensor</dt>
	<dd>Détecteur d'ouverture</dd>
</dl>
<li>Divers :</li>
<dl>
	<dt>HC_user :</dt>
	<dd>Utilisateur de la box</dd> 
	<dt>weather :</dt>
	<dd>Informations météo</dd>
	<dt>iOS_device</dt>
	<dd>Téléphone ou tablette iOS</dd>
	<dt>virtual_device</dt>
	<dd>Périphérique virtuel</dd>
</dl>
</ul>

<h2>3. ACTIONS : <tt>homecenter.xml</tt></h2>

<p>Ce fichier contient l'ensemble de la grammaire et les actions à réaliser.<br />
Quelques exemples :</p>
<pre>
	<font color="blue">
	&lt;one-of&gt;
		&lt;item&gt;ouvre		&lt;tag&gt;out.action.actionModule="turnOn";	out.action.ttsAction="j'ouvre";	out.action.button="1";	&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;ferme		&lt;tag&gt;out.action.actionModule="turnOff";	out.action.ttsAction="je ferme";out.action.button="3";	&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;allume		&lt;tag&gt;out.action.actionModule="turnOn";	out.action.ttsAction="j'allume";	&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;éteint		&lt;tag&gt;out.action.actionModule="turnOff";	out.action.ttsAction="j'éteint";	&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;ferme		&lt;tag&gt;out.action.actionModule="turnOff";	out.action.ttsAction="je ferme";	&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;augmente		&lt;tag&gt;out.action.actionModule="setValue";out.action.ttsAction="oké, je règle";	&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;quelle est	&lt;tag&gt;out.action.actionModule="getValue";out.action.ttsAction="";		&lt;/tag&gt;&lt;/item&gt;
	&lt;/one-of&gt; 
	</font><font color="green">
	&lt;one-of&gt;
		&lt;!-- VOLETS --&gt;
		&lt;item&gt;le volet de la salle a manger	&lt;tag&gt;out.action.module="volet_sam";&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;le volet du séjour		&lt;tag&gt;out.action.module="volet_sejour";&lt;/tag&gt;&lt;/item&gt;
		&lt;!-- LAMPES --&gt;
		&lt;item&gt;la cuisine			&lt;tag&gt;out.action.module="lampe_cuisine";&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;la lampe de la cuisine		&lt;tag&gt;out.action.module="lampe_cuisine";&lt;/tag&gt;&lt;/item&gt;
		&lt;item&gt;la lumière de la cuisine		&lt;tag&gt;out.action.module="lampe_cuisine";&lt;/tag&gt;&lt;/item&gt;
		&lt;!-- Sondes --&gt;
		&lt;item&gt;la température du salon		&lt;tag&gt;out.action.module="temp_salon";&lt;/tag&gt;&lt;/item&gt;	
		&lt;item&gt;l'hygrométrie de la salle de bain	&lt;tag&gt;out.action.module="hygro_sdb";&lt;/tag&gt;&lt;/item&gt;	
		&lt;item&gt;la luminosité de l'entrée		&lt;tag&gt;out.action.module="lumi_entree";&lt;/tag&gt;&lt;/item&gt;	
	&lt;/one-of&gt;
	</font><font color="red">
	&lt;item repeat="0-1"&gt;
		&lt;one-of&gt;
			&lt;item&gt;à vingt cinq pour cent		&lt;tag&gt;out.action.ttsDim=" à vingt cinq pour cent";	out.action.setValue=25;&lt;/tag&gt;&lt;/item&gt;
			&lt;item&gt;à cinquante pour cent		&lt;tag&gt;out.action.ttsDim=" à cinquante pour cent";	out.action.setValue=50;&lt;/tag&gt;&lt;/item&gt;
			&lt;item&gt;à soixante quinze pour cent	&lt;tag&gt;out.action.ttsDim=" à soixante quinze pour cent";	out.action.setValue=75;&lt;/tag&gt;&lt;/item&gt;
			&lt;item&gt;à cent pour cent			&lt;tag&gt;out.action.ttsDim=" à cent pour cent";		out.action.setValue=100;&lt;/tag&gt;&lt;/item&gt;
		&lt;/one-of&gt;
	&lt;/item&gt;
	</font>
</pre>

<p>Les informations importantes de ce fichier sont :</p>

<ul>
<li><b>actionModule</b></li>	
<p>Action à réaliser ; les choix possibles sont : 
	<ul>
		<li>Commandes de type tout-ou-rien (<tt>turnOn</tt> et <tt>turnOff</tt>) :</li>  
			<p>Utilisée pour allumer ou éteindre les lumières, activer ou désactiver les prises commandées, ouvrir ou fermer les volets roulants, etc.</p>
			<p>Pour les périphériques à sortie variable tels que les gradateurs ou les commandes de volets roulants,
			et si la commande comporte une information de niveau (exemple : <i>« à 50 % »</i>) 
			ce paramètre est ignoré au bénéfice de <tt>setValue</tt> (cf ci-dessous).</p>
		<li>Commande de réglage (<tt>setValue</tt>) :</li> 
			<p>Utilisée pour régler l'intensité d'une lampe ou la position d'un volet roulant.</p>
			<p>Si vous omettez de préciser le réglage, cette action sera remplacée par <tt>turnOn</tt>.</p>
		<li>Commande de lecture (<tt>getValue</tt>) :</li> 
			<p>Utilisé pour lire la donnée d'une sonde, l'état d'un module, connaître la météo, etc.</p>
	</ul>
	
<li><b>ttsAction</b></li>      
<p>Phrase prononcée par SARAH lorsque l'action est réalisée.</p>

<li><b>module</b></li>         
<p>Le nom du module concerné par l'action. Ce nom doit correspondre à une (et une seule) des lignes <tt>&lt;device&gt;</tt> du fichier <tt>devices.xml</tt>.</p>

<li><b>setValue</b></li>
<p>Valeur à transmettre au module à sortie variable (dimmer, volet-roulant, etc.).<br />
Cette valeur est comprise en 0 et 100 %.</p>

<li><b>ttsDim</b></li>
<p>Phrase prononcée par SARAH pour confirmer le réglage.</p>
</ul>

<h2>4. COMMENT ÇA MARCHE ?</tt></h2>

<p>Dans le code de la section précédente, 3 blocs sont mis en évidence. 
Nous allons les illustrer avec la commande suivante :<br />
« <i> Sarah, allume la lampe de la cuisine à 50 %</i> ».</p>
<ul>
	<li>La description des <b>ordres</b> (en bleu)</li>
		<p>La première partie, juste après la balise <tt>&lt;item&gt;</tt>, 
		détaille la phrase à prononcer (toujours précédée de "Sarah", pour rappel).<br />
		Exemple : <i>Sarah, allume ...</i> va activer la ligne suivante :</p>
		<font color="blue">
		<pre>&lt;item&gt;allume		&lt;tag&gt;out.action.actionModule="turnOn";	out.action.ttsAction="j'allume";&lt;/tag&gt;&lt;/item&gt;</pre>
		</font>
		<p>À ce stade, SARAH sait qu'il faut allumer (car <tt>actionModule="turnOn"</tt>) 
		un module, mais ne sait pas encore lequel. 
		Elle mémorise également le début de sa réponse <tt>tts</tt> 
		(grâce à <tt>ttsAction=</tt><i>"j'allume"</i>)</p>


changelog.md
Version 0.3.5 (29/08/2016) <br />
— added rollershutter support

Version 0.3.4 (27/01/2016) <br />
— added 'update' command to automagically grab every module declared in the homecenter

Version 0.3.3 (26/01/2016) <br />
— changes in XML parameters

Version 0.3.2 (26/01/2016) <br />
— forced turnOn if setValue undefined<br />
— modified console logs<br />
— code cleaning <br />

Version 0.3.1 (25/01/2016) <br />
— bugfix : replaced body.properties.value with body.properties['ui.VoletDIM.value'] in virtual_device<br />

Version 0.3.0 (25/01/2016) <br />
— support du plugin Logitech Harmony<br />

Version 0.2.1 (24/01/2016) <br />
— support des modules virtuels : sliders<br />

Version 0.2.0 (23/01/2016) <br />
— support des modules virtuels : boutons<br />

Version 0.1.2 (22/01/2016) <br />
— support du module météo<br />

Version 0.1.1 (21/01/2016) <br />
— support des sondes<br />

Version 0.1.0 (20/01/2016) <br />
— support des binarySwitch multilevelSwitch<br />


plugins/homecenter/homecenter.js
/*
Version 0.3.5 (29/08/2016)
— added rollershutter support
Version 0.3.4 (27/01/2016)
— added 'update' command to automagically grab every module declared in the homecenter
Version 0.3.3 (26/01/2016)
— changes in XML parameters
Version 0.3.2 (26/01/2016)
— forced turnOn if setValue undefined
— modified console logs
— code cleaning
Version 0.3.1 (25/01/2016)
— bugfix : replaced body.properties.value with body.properties['ui.VoletDIM.value'] in virtual_device
Version 0.3.0 (25/01/2016)
— support du plugin Logitech Harmony
Version 0.2.1 (24/01/2016)
— support des modules virtuels : sliders
Version 0.2.0 (23/01/2016)
— support des modules virtuels : boutons
Version 0.1.2 (22/01/2016)
— support du module météo
Version 0.1.1 (21/01/2016)
— support des sondes
Version 0.1.0 (20/01/2016)
— support des binarySwitch multilevelSwitch
*/

exports.action = function(data, callback, config) 
{
	var xmldoc = require('./lib/xmldoc');
	config = config.modules.homecenter;
	var actionList 		= ['turnOff', 'turnOn', 'setValue', 'getValue'];
	var api_devices 	= '/api/devices/';
	var api_weather 	= '/api/weather/';
	var action_path 	= '/action/';
	var http 			= 'http://';
	var url 			= '';
	var json_body_start	= '{"args": [';
	var json_body_null 	= 'null';
	var json_body_end 	= ']}';
	var json_body_value = '';
	var json_body 		= '';
	
	/** Module access **/
	var fs = require('fs');
	var xmlFile = fs.readFileSync(__dirname + '\\devices.xml');
	var file = new xmldoc.XmlDocument(xmlFile);
	var module = file.childWithAttribute('name', data.module);

	if (!config.url) {
		console.log("|HC|Error: Missing Home Center url");
		return;
	}
	// 
	// Devices action : construct URL
	// 
	var reqmethod = 'POST';
	console.log("|HC|Info.: Action = " + data.command);
	switch(data.command) {
	case 'turnOn' 	: 
	case 'turnOff' 	:
	case 'setValue'	:
		url = http + config.url + ":" + config.port + api_devices + module.attr.id + action_path;
		console.log("|HC|Info.: Type = " + module.attr.type);
		switch(module.attr.type) {
		case 'com.fibaro.multilevelSwitch' :
		case 'com.fibaro.blind' :		
			if (data.dimValue) {
				if (module.attr.type == 'multilevelSwitch') {
					console.log("|HC|Info.: Dimmer set to " + data.dimValue);
				} else {
					console.log("|HC|Info.: Rollershutter set to " + data.dimValue);				
				}
				url += 'setValue';
				json_body_value = data.dimValue;
				break;
			} else {
				if (module.attr.type == 'multilevelSwitch') {
					console.log("|HC|Info.: Dimmer value not set : turning ON or OFF");
				} else {
					console.log("|HC|Info.: Rollershutter value not set : turning ON or OFF");				
				}
				// Switching to next case
			}
		case 'com.fibaro.binarySwitch' :
			if (data.command == 'setValue') data.command = 'turnOn';
			url += data.command;
			json_body_value = json_body_null;
			break;
		case 'com.fibaro.logitechHarmonyActivity' :
			switch (data.command) { 
			case 'turnOn':
				url += 'changeActivityState';
				break;
			case 'turnOff':
				url += 'changeActivityState';
				break;
			default :
				console.log("|HC|Warn.: You shouldn't have reached this point !");
				console.log("|HC|Warn.: => Please verify 'homecenter.js' file !");
				return callback({'tts': "Désolée, erreur dans le traitement de la requête pour le module de type " + module.attr.type});
			}
			break;
		case 'virtual_device' :
// Pour volets Somfy RTS via passerelle Zibase et VD de Zorgonaute84 
// http://www.domotique-fibaro.fr/index.php/topic/3067-module-virtuel-pour-volet-somfy-avec-slider 
			if (data.dimValue) {
				 url += 'setSlider';
				json_body_value = module.attr.slider + "," + data.dimValue;
			}
			else {
				switch (data.command) { 
				case 'turnOn':
					url += 'pressButton';
					json_body_value = module.attr.buttonOn;
					break;
				case 'turnOff':
					url += 'pressButton';
					json_body_value = module.attr.buttonOff;
					break;
				case 'setValue':
					url += 'pressButton';
					json_body_value = module.attr.buttonMy;
				default :
					console.log("|HC|Warn.: You shouldn't have reached this point !");
					console.log("|HC|Warn.: => Please verify 'homecenter.js' file !");
					return callback({'tts': "Désolée, erreur dans le traitement de la requête pour le module de type " + module.attr.type});
				}
			}
			break;
		case 'com.fibaro.lightSensor' 		:
		case 'com.fibaro.temperatureSensor' :
		case 'com.fibaro.humiditySensor' 	:
		case 'com.fibaro.seismometer' 		:
		case 'com.fibaro.multilevelSensor' 	: 
		case 'com.fibaro.motionSensor'		:
		case 'com.fibaro.FGMS001'			:
		case 'com.fibaro.FGFS101'			:
		case 'com.fibaro.FGSS001'			:
		case 'com.fibaro.heatDetector'		:
		case 'com.fibaro.doorSensor'		:
		case 'weather' 						:
			console.log("|HC|Warn.: Device " + module.attr.type + " not designed for POST method");
			return callback({'tts': "Désolée, les modules " + module.attr.type + " n'acceptent pas ce type de commande"});
			break;
		default :
			console.log("|HC|Warn.: Unknown device '" + module.attr.type + "'!")
			console.log("|HC|Warn.: Check your 'devices.xml' file");
			return callback({'tts': "Désolée, le type de module " + module.attr.type + "m'est inconnu"});
		}
		break;
	case 'getValue' :
		reqmethod='GET';
		json_body='';
		console.log("|HC|Info.: Type = " + module.attr.type);
		switch(module.attr.type) {
		case 'com.fibaro.multilevelSwitch' 	:
		case 'com.fibaro.blind' 			:
		case 'com.fibaro.binarySwitch' 		:
		case 'com.fibaro.logitechHarmonyActivity' :
		case 'com.fibaro.lightSensor' 		:
		case 'com.fibaro.temperatureSensor' :
		case 'com.fibaro.humiditySensor' 	:
		case 'com.fibaro.seismometer' 		:
		case 'com.fibaro.multilevelSensor' 	: 
			url = http + config.url + ":" + config.port + api_devices + module.attr.id;
			break;
		case 'weather' : // Weather action
			url = http + config.url + ":" + config.port + api_weather;
			break;
		case 'virtual_device' :
			url = http + config.url + ":" + config.port + api_devices + module.attr.id;
			break;
		case 'HC_user':
		case 'iOS_device':
		case 'com.fibaro.motionSensor'		:
		case 'com.fibaro.FGMS001'			:
		case 'com.fibaro.FGFS101'			:
		case 'com.fibaro.FGSS001'			:
		case 'com.fibaro.heatDetector'		:
		case 'com.fibaro.doorSensor'		:
			console.log("|HC|Warn.: Module type '" + module.attr.type + "' not yet implemented.");
			return callback({'tts': "Désolée, les modules de type " + module.attr.type + "ne sont pas encore implémentés."});
		default : // Scenario action
			console.log("|HC|Warn.: Unknown device '" + module.attr.type + "'!")
			console.log("|HC|Warn.: Check your 'devices.xml' file");
			return callback({'tts': "Désolée, le type de module " + module.attr.type + "m'est inconnu"});
		}	
		break;
	case 'update':
		if ( (module.attr.name == 'modules') && (module.attr.type == 'update') ) {
			reqmethod='GET';
			json_body='';
			url = 'http://' + config.url + ":" + config.port + '/api/devices';
		} else {
			console.log("|HC|Warn.: Command '" + data.command + "' didn't comply to double security check !");
			console.log("|HC|Warn.: You can only launch an update by asking it for a  module named ");
			console.log("|HC|Warn.: 'modules' and of 'update' type. Check your 'devices.xml' file.");
			return callback({'tts': "Désolée, cette commande ne sert qu'à mettre à jour tous les modules. Ne l'utilisez pas à la légère."});
		}
		break;
	default :
		console.log("|HC|Warn.: Unknown command '" + data.command + "'!");
		console.log("|HC|Warn.: Check your 'homecenter.xml' file");
		return callback({'tts': "Désolée, la commande " + data.command + "m'est inconnue"});
	}
	
	console.log("|HC|Info.: Sending request to: ");
	console.log("|HC|Info.: " + url);
	json_body = json_body_start + json_body_value + json_body_end;

	//	
	// Send Request
	//

	var request = require('request');
	var options = {
		method: reqmethod,
		uri: url,
		body: json_body,
		headers: {
    		'Authorization': 'Basic ' + new Buffer(config.username+":"+config.password).toString('base64')
		}
	};
	var tts_sensor_value = "";
	request(options, function(error, response, body) {
    	if (!error && (response.statusCode == 200) || (response.statusCode == 202)) {
        	body=JSON.parse(body);
     	}
     	else {
     		console.log("|HC|Error: Error with url: " + url);
     	}
		var tts = data.ttsCmd + " " + module.attr.tts;
		
		console.log("|HC|Info.: Id = " + module.attr.id);
		console.log("|HC|Info.: Name = " + module.attr.name);
	
	//
	// Parse JSON answer for GET method
	//
			
		switch (data.command) {
		case 'getValue' :
			switch(module.attr.type) {
			case 'HC_user':
			case 'iOS_device':
			case 'com.fibaro.motionSensor'		:
			case 'com.fibaro.FGMS001'			:
			case 'com.fibaro.FGFS101'			:
			case 'com.fibaro.FGSS001'			:
			case 'com.fibaro.heatDetector'		:
			case 'com.fibaro.doorSensor'		:
				console.log("|HC|Warn.: Module type '" + module.attr.type + "' not yet implemented.");
				return callback({'tts': "Désolée, les modules de type " + module.attr.type + "ne sont pas encore implémentés."});
			case 'weather':
				var weather_humidity 	= parseFloat(body.Humidity).toString().replace("."," virgule ");
				var weather_wind 		= parseFloat(body.Wind).toString().replace("."," virgule ");
				var weather_temperature = parseFloat(body.Temperature).toString().replace("."," virgule ");
				weather_temperature = weather_temperature.replace("-"," moins ")
				tts += "La température est de " 	+ weather_temperature + " degrés. ";
				if (parseFloat(body.Temperature) < 0) tts += "Attention au risque de gel. ";
				tts += "L'hygrométrie est de " 		+ weather_humidity + " pour cent. ";
				tts += "La vitesse du vent est de "	+ weather_wind + " kilomètres heure. ";
				break;
			case 'virtual_device':
				switch (parseFloat(body.properties['ui.VoletDIM.value'])) { 
					case 0   : tts += " est ouvert"; break;
					case 99  : 
					case 100 : tts += " est fermé"; break;
					default  : tts += " est fermé à " + body.properties['ui.VoletDIM.value'] + " pour cent";
				}
				break;
			case 'com.fibaro.logitechHarmonyActivity' :
				switch (body.properties['ui.activityCurrentStateValueLabel.caption']) {
				case 'Activity OFF' :
					tts += " est arrêté.";
					break;
				case 'Running' :
					tts += " est en marche.";
					break;
				default :
					tts += " est dans un état inconnu.";
				}
				break;
			case 'com.fibaro.multilevelSwitch':
				switch (parseFloat(body.properties.value)) { 
				case 0   : tts += " est éteinte"; break;
				case 99  : 
				case 100 : tts += " est allumée"; break;
				default  : tts += " est allumée à " + body.properties.value + " pour cent";
				}
				break;
			case 'com.fibaro.blind':
				switch (parseFloat(body.properties.value)) { 
				case 0   : tts += " est fermé"; break;
				case 99  : 
				case 100 : tts += " est ouvert"; break;
				default  : tts += " est ouvert à " + body.properties.value + " pour cent";
				}
				break;
			case 'com.fibaro.binarySwitch':
				switch (body.properties.value) {
				case 'true' : tts += " est allumée"; break;
				case 'false': tts += " est éteinte"; break;
				default     : tts += " est dans un état inconnu";
				}
				break;
			case 'com.fibaro.temperatureSensor'	:
   			case 'com.fibaro.humiditySensor'	:
       		case 'com.fibaro.lightSensor'		:
        	case 'com.fibaro.seismometer'		:
   	    	case 'com.fibaro.multilevelSensor'	:
				tts_sensor_value = parseFloat(body.properties.value).toString().replace("."," virgule ");
				break;
	  		default:
				console.log("|HC|Warn.: Unknown sensor '" + module.attr.type + "'!");
				console.log("|HC|Warn.: Check your 'device.xml' file");
			}
			break;
		case 'update':
			var device_type = [	
				// Commentez les types que vous ne souhaitez pas utiliser
				  'com.fibaro.binarySwitch'
				, 'com.fibaro.multilevelSwitch'
				, 'com.fibaro.blind'
				, 'com.fibaro.temperatureSensor'
				, 'com.fibaro.humiditySensor'
				, 'com.fibaro.lightSensor'
				, 'com.fibaro.multilevelSensor'
				, 'weather'
				//, 'com.fibaro.motionSensor'
				//, 'com.fibaro.FGFS101'
				//, 'com.fibaro.FGSS001'
				//, 'com.fibaro.heatDetector'
				//, 'com.fibaro.doorSensor'
				, 'virtual_device'
				, 'com.fibaro.logitechHarmonyActivity'
				//, 'HC_user'	
				//, 'iOS_device'
			];

			console.log("|HC|Info.: ##### UPDATING XML FILES #####");
			var now = Date();
			var fs   = require('fs');
			var grmfile 	= './plugins/homecenter/homecenter.xml';
			var grmxml  	= fs.readFileSync(grmfile,'utf8');
			var grmregexp 	= new RegExp('@[^@]+@','gm');
			var grmstring	= [];
			//console.log("|HC|Info.:Saving old 'homecenter.xml' file to 'homecenter-" + dateFormat(now, "yyyymmdd-HHmmss") + ".xml'");
			//fs.writeFileSync('homecenter-'+dateFormat(now, "yyyymmdd-HHmmss")+'.xml', grmxml, 'utf8');
			console.log("|HC|Info.:Saving old 'homecenter.xml' file to 'homecenter-.xml'");
			fs.writeFileSync('./plugins/homecenter/homecenter-.xml', grmxml, 'utf8');
			var devfile  	= './plugins/homecenter/devices.xml';
			var devxml   	= fs.readFileSync(devfile,'utf8');
			var devstring 	= [];
			var devregexp 	= new RegExp('@[^@]+@','gm');
			//console.log("|HC|Info.:Saving old 'devices.xml' file to 'devices-" + dateFormat(now, "yyyymmdd-HHmmss") + ".xml'");
			//fs.writeFileSync('devices-'+dateFormat(now, "yyyymmdd-HHmmss")+'.xml', devxml, 'utf8');
			console.log("|HC|Info.:Saving old 'devices.xml' file to 'devices-.xml'");
			fs.writeFileSync('./plugins/homecenter/devices-.xml', devxml, 'utf8');

			//
			// Looking for devices types, names, and ids
			//
			for ( var i = 0; i < body.length; i++ ) {
				if (body[i].visible == true) {
					if (grmstring[body[i].type])
						grmstring[body[i].type] += '\t\t<item>' + body[i].name + '\t\t<tag>out.action.module="' + body[i].name + '"</tag></item>\r\n';
					else grmstring[body[i].type] = '\t\t<item>' + body[i].name + '\t\t<tag>out.action.module="' + body[i].name + '"</tag></item>\r\n';
					if (devstring[body[i].type]) 
						devstring[body[i].type]  += '\t<device name="' + body[i].name + '" \t\ttype="' + body[i].type + '" \tid="' + body[i].id + '" \ttts="' + body[i].name + '" />\r\n';
					else devstring[body[i].type]  = '\t<device name="' + body[i].name + '" \t\ttype="' + body[i].type + '" \tid="' + body[i].id + '" \ttts="' + body[i].name + '" />\r\n';
				}
			}
			//
			// Formatting 'homecenter.xml' file content
			//
			var grmcontent 	= '@ -->\r\n' 
							+ '<!--=== FAITES VOS MODIFICATIONS SOUS CETTE LIGNE UNIQUEMENT ====-->\r\n';
			for (var i = 0 ; i < device_type.length ; i++) { 
				if ( grmstring[device_type[i]] ) {
					grmcontent += '\r\n\t\t<!-- Modules de type ' +  device_type[i] + ' -->\r\n\r\n';
					grmcontent += grmstring[device_type[i]];
				}
			}
			grmcontent 	+= '<!--============= NE PLUS MODIFIER SOUS CETTE LIGNE =============-->\r\n' 
						+  '<!-- @';
			grmxml = grmxml.replace(grmregexp,grmcontent);
			fs.writeFileSync(grmfile, grmxml, 'utf8');
			//
			// Formatting 'devices.xml' file content
			//
			var devcontent 	= '@ -->\r\n' 
							+ '<!--=== FAITES VOS MODIFICATIONS SOUS CETTE LIGNE UNIQUEMENT ====-->\r\n';
			for (var i = 0 ; i < device_type.length ; i++) { 
				if ( devstring[device_type[i]] ) {
					devcontent += '\r\n\t<!-- Modules de type ' +  device_type[i] + ' -->\r\n\r\n';
					devcontent += devstring[device_type[i]];
				}
			}
			devcontent  += '<!--============= NE PLUS MODIFIER SOUS CETTE LIGNE =============-->\r\n\r\n' 
						+ '<!-- @';
			devxml = devxml.replace(devregexp,devcontent);
			fs.writeFileSync(devfile, devxml, 'utf8');
			//
			// Quit
			//
			console.log("|HC|Info.: ##### UPDATE SUCCESSFULL #####");
			//SARAH.speak("Mise à jour terminée. Vérifiez vos fichiers de configuration pour l'adapter à vos besoins.");
			tts = "Mise à jour terminée. Vérifiez vos fichiers de configuration pour l'adapter à vos besoins.";
			break;
			
		default:
			tts_sensor_value = "";
		}

		//
		// tts
		//
		
		tts = tts.replace('%s', tts_sensor_value);
		if (data.ttsDim){
			tts += " " + data.ttsDim;
		}
		console.log("|HC|Speak: " + tts);
		// Callback with TTS
		callback({'tts': tts});
	});
}
